File: casciffo-spring-backend/src/main/kotlin/isel/casciffo/casciffospringbackend/research/ResearchModel.kt
package isel.casciffo.casciffospringbackend.research

import isel.casciffo.casciffospringbackend.proposals.ProposalModel
import isel.casciffo.casciffospringbackend.proposals.ResearchType
import isel.casciffo.casciffospringbackend.research.patients.Participant
import isel.casciffo.casciffospringbackend.states.State
import isel.casciffo.casciffospringbackend.states.transitions.StateTransition
import org.springframework.beans.factory.annotation.Value
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import reactor.core.publisher.Flux
import java.time.LocalDate

@Table("clinical_research")
data class ResearchModel (
    @Id
    @Column("research_id")
    var id: Int? = null,

    var proposalId: Int? = null,

    @Column("research_state_id")
    var stateId: Int? = null,

    var eudra_ct: String? = null,

    var sampleSize: Int? = null,

    var duration: Int? = null,

    var cro: String? = null,

    var startDate: LocalDate? = null,

    var endDate: LocalDate? = null,

    var estimatedEndDate: LocalDate? = null,

    var industry: String? = null,

    var protocol: String? = null,

    var initiativeBy: String? = null,

    var phase: String? = null,

    var type: ResearchType? = null,

    @Transient
    @Value("null")
    var state: State? = null,

    @Transient
    @Value("null")
    var proposal: ProposalModel? = null,

    @Transient
    @Value("null")
    var participants: Flux<Participant>? = null,

    @Transient
    @Value("null")
    var stateTransitions: Flux<StateTransition>? = null
)

File: demo/front-end-pwa/src/model/research/ResearchModel.ts
import {ProposalModel} from "../proposal/ProposalModel";
import {StateTransitionModel} from "../state/StateTransitionModel";
import {StateModel} from "../state/StateModel";
import UserModel from "../user/UserModel";
import {TeamInvestigatorModel} from "../TeamInvestigatorModel";


export interface ResearchModel {
    id?: string
    eudra_ct?: string
    sampleSize?: string
    duration?: string
    cro?: string
    startDate?: string
    endDate?: string
    estimatedEndDate?: string
    estimatedPatientPool?: string
    actualPatientPool?: string
    industry?: string
    protocol?: string
    initiativeBy?: string
    phase?: string
    type?: string
    treatmentType?: string
    typology?: string
    specification?: string
    stateId?: string
    state?: StateModel
    proposalId?: string
    proposal?: ProposalModel
    visits?: ResearchVisitModel[]
    stateTransitions?: StateTransitionModel[]
    dossiers?: DossierModel[]
    patients?: PatientModel[]
    scientificActivities?: ScientificActivityModel[]
    investigationTeam?: TeamInvestigatorModel[]
}

export interface ResearchAggregateModel {
    id?: string
    eudra_ct?: string
    sampleSize?: string
    duration?: string
    cro?: string
    startDate?: string
    endDate?: string
    estimatedEndDate?: string
    industry?: string
    protocol?: string
    initiativeBy?: string
    phase?: string
    type?: string
    stateId?: string
    stateName?: string
    proposalId?: string
    sigla?: string
    serviceName?: string
    serviceId?: string
    therapeuticAreaName?: string
    therapeuticAreaId?: string
    pathologyName?: string
    pathologyId?: string
    principalInvestigatorId?: string
    principalInvestigatorName?: string
    principalInvestigatorEmail?: string
    promoterName?: string
}

export interface DossierModel {
    id?: number
    clinicalResearchId?: string
    volume?: string
    label?: string
    amount?: number
}

export interface ResearchVisitModel {
    id?: string
    researchId?: string
    participantId?: string
    concluded?: boolean;
    visitType?: string
    scheduledDate?: string
    startDate?: string
    endDate?: string
    periodicity?: string
    observations?: string
    hasAdverseEventAlert?: boolean
    hasMarkedAttendance?: boolean
    patient?: PatientModel
    visitInvestigators?: VisitInvestigator[]
}

export interface VisitInvestigator {
    id?: string
    visitId?: string
    investigatorId?: string
    investigator?: UserModel
}

export interface PatientModel {
    id?: string
    processId?: string
    fullName?: string
    gender?: string
    age?: string
    joinDate?: string
    lastVisitDate?: string
    treatmentBranch?: string
}

export interface PatientVisitsAggregate {
    patient: PatientModel
    scheduledVisits: ResearchVisitModel[]
}

export interface ScientificActivityModel {
    id?: string
    researchId?: string
    datePublished?: string
    author?: string
    paperName?: string
    volume?: string
    volumeNumber?: string
    paperNumPages?: string
    countryPublished?: string
    hasBeenIndexed?: boolean
    publishedUrl?: string
    publicationType?: string
    studyType?: string
}
